src/prefect/task_engine.py

```diff
@@ -368,34 +371,64 @@ def key_fn():
         self.set_state(terminal_state)
         return result
 
-    def handle_retry(self, exc: Exception) -> bool:
-        """
-        If the task has retries left, and the retry condition is met, set the task to retrying.
+    @sync_compatible
```

Comment:
Still thinking through broader recommendations for unifying sync/async code, but I don't like @sync_compatible on internal routes. It makes it very difficult to understand what happens when you look at the code. I'd prefer making explict async methods and then calling them with `run_coro_as_sync` when they are needed in a sync frame. This is more explicit and makes it very clear what happens and where a function will be run. @sync_compatible should really only be used when we need to expose a single function that is user-facing (e.g. Block.load) such that we can't know in advance how it's going to be used (and even then I think is more destructive than helpful)